tree/dataframe/src/RDFSnapshotHelpers.cxx

```diff
@@ -930,3 +995,260 @@ ROOT::Internal::RDF::UntypedSnapshotRNTupleHelper::MakeNew(void *newName)
+/// of Snapshot with systematic uncertainties.
+struct SnapshotOutputWriter {
+   std::unique_ptr<TFile> fFile;
+   std::unique_ptr<TTree> fTree;
+   std::string fDirectoryName;
+   RLoopManager *fOutputLoopManager;
+
+   // Bitmasks to indicate whether syst. uncertainties have been computed. Bound to TBranches, so need to be stable in
+   // memory.
+   struct Bitmask {
+      std::string branchName;
+      std::bitset<64> bitset{};
+      std::unique_ptr<uint64_t> branchBuffer{new uint64_t{}};
+   };
+   std::vector<Bitmask> fBitMasks;
+
+   std::unordered_map<std::string, unsigned int> fBranchToVariationMapping;
+   // The corresponding ROOT dictionary is declared in core/clingutils/src
+   std::unordered_map<std::string, std::pair<std::string, unsigned int>> fBranchToBitmaskMapping;
+   unsigned int fNBits = 0;
+
+   SnapshotOutputWriter(TFile *file, TTree *tree) : fFile{file}, fTree{tree} {}
+   ~SnapshotOutputWriter()
+   {
+      if (!fBranchToBitmaskMapping.empty()) {
+         fFile->WriteObject(&fBranchToBitmaskMapping,
+                            (std::string{"R_rdf_branchToBitmaskMapping_"} + fTree->GetName()).c_str());
+      }
+      if (fTree) {
+         // use AutoSave to flush TTree contents because TTree::Write writes in gDirectory, not in fDirectory
+         fTree->AutoSave("flushbaskets");
+
+         // Now connect the data source to the loop manager so it can be used for further processing
+         std::string tree = fTree->GetName();
+         if (!fDirectoryName.empty())
+            tree = fDirectoryName + '/' + tree;
+         std::string file = fFile->GetName();
+
+         fTree.reset();
+         fFile.reset();
+
+         if (fOutputLoopManager)
+            fOutputLoopManager->SetDataSource(std::make_unique<ROOT::Internal::RDF::RTTreeDS>(tree, file));
+      }
+   }
+   SnapshotOutputWriter(SnapshotOutputWriter const &) = delete; // Anyway deleted because of the unique_ptrs
+   SnapshotOutputWriter &operator=(SnapshotOutputWriter const &) = delete;
+   SnapshotOutputWriter(SnapshotOutputWriter &&) noexcept =
+      delete; // Can be done, but need to make move-from object safe to destruct
+   SnapshotOutputWriter &operator=(SnapshotOutputWriter &&) noexcept = delete;
+
+   /// Register a branch and corresponding systematic uncertainty.
+   /// This will create an entry in the mapping from branch names to bitmasks, so the corresponding
+   /// column can be masked if it doesn't contain valid entries. This mapping is written next to the
+   /// tree into the output file.
+   void RegisterBranch(std::string const &branchName, unsigned int variationIndex)
+   {
+      if (auto it = fBranchToVariationMapping.find(branchName); it != fBranchToVariationMapping.end()) {
+         if (variationIndex != it->second) {
+            throw std::logic_error("Branch " + branchName + " is being registered with different variation index.");
```

Comment:
Maybe add the following (or maybe consider also writing the full name of the variation in the error message?)
```suggestion
            throw std::logic_error("Branch " + branchName + " is being registered with different variation index than the expected one: " + std::to_string(variationIndex));
```